### PR TITLE
Sync calls

### PR DESCRIPTION
**Pull request by abizjak**
## Purpose

Addresses #71 

## Changes

- Remove actions from the concordium-std library, and wrap the "invoke" functions in a type-safe and high-level interface
- Update procedural macros to generate entrypoints compatible with the new V1 contract API
- Update testing infrastructure

**Remaining tasks**
- [x] Polish
- [x] Introduce V2 version of the schema that supports return values (but no longer contract state)
- [x] Update documentation
- [ ] Update the example contracts **(postponed until the new state PR)**

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: concordium-std/src/test_infrastructure.rs [hunk 1]
```diff
@@ -695,6 +649,223 @@ impl<T: AsRef<[u8]>> Seek for ContractStateTest<T> {
...
+        if *contract_balance >= amount {
+            *contract_balance -= amount;
+            self.transfers.push((receiver.clone(), amount));
+            Ok(())
+        } else {
+            Err(InvokeError::AmountTooLarge)
+        }
+    }
+
+    /// Invoke a contract.
```
Comment (Bargsteen) [inline concordium-std/src/test_infrastructure.rs:745]: Docs

File: concordium-std/src/test_infrastructure.rs [hunk 2]
```diff
@@ -695,6 +649,223 @@ impl<T: AsRef<[u8]>> Seek for ContractStateTest<T> {
...
+        let mut output = Vec::new();
+        let res = (handler.mock_fn)(parameter, amount, &mut self.state, &mut output)
+            .and_then(|state_modified| Ok((state_modified, Some(Cursor::new(output)))));
+        if res.is_ok() && amount.micro_ccd > 0 {
+            *unwrap_contract_balance(&mut self.contract_balance) -= amount;
+        }
+        res
+    }
+
+    fn state(&mut self) -> &mut State { &mut self.state }
```
Comment (Bargsteen) [inline concordium-std/src/test_infrastructure.rs:775]: Docs

File: concordium-std/src/test_infrastructure.rs [hunk 3]
```diff
@@ -695,6 +649,223 @@ impl<T: AsRef<[u8]>> Seek for ContractStateTest<T> {
...
+
+    fn self_balance(&self) -> Amount {
```
Comment (Bargsteen) [inline concordium-std/src/test_infrastructure.rs:777]: Docs

File: concordium-std/src/test_infrastructure.rs [hunk 4]
```diff
@@ -695,6 +649,223 @@ impl<T: AsRef<[u8]>> Seek for ContractStateTest<T> {
...
+                method
+            ),
+        };
+        // Check if the contract has sufficient balance.
+        // Do not try to unwrap the balance if amount is zero.
+        if amount.micro_ccd > 0 && *unwrap_contract_balance(&mut self.contract_balance) < amount {
+            return Err(InvokeError::AmountTooLarge);
+        }
+        let mut output = Vec::new();
+        let res = (handler.mock_fn)(parameter, amount, &mut self.state, &mut output)
```
Comment (Bargsteen) [inline concordium-std/src/test_infrastructure.rs:767]: Docs

File: concordium-std/src/test_infrastructure.rs (hunk 3 above)
Comment (Bargsteen) [reply concordium-std/src/test_infrastructure.rs:777]: Add comment about why this doesn't use unwrap_contract_balance (not possible due to immutable ref to self)

File: concordium-std/src/test_infrastructure.rs [hunk 5]
```diff
@@ -695,6 +649,223 @@ impl<T: AsRef<[u8]>> Seek for ContractStateTest<T> {
...
+        Some(ref mut amount) => amount,
+        None => fail!(
+            "Unset contract balance on the test host. Make sure to use `set_balance` if your \
+             contract uses the balance."
+        ),
+    }
+}
+
+impl<State> HostTest<State> {
+    pub fn new(state: State) -> Self {
```
Comment (Bargsteen) [inline concordium-std/src/test_infrastructure.rs:801]: Docs for all of the methods

**Review by Bargsteen — CHANGES_REQUESTED** [review]

File: concordium-std/src/impls.rs [hunk 6]
```diff
@@ -419,12 +473,145 @@ impl<T: sealed::ContextType> HasCommonData for ExternContext<T> {
...
+        // serialize state since it might have been modified
+        // FIXME: Only do this if needed.
+        self.state.serial(&mut ContractState::open(())).unwrap_abort();
+        let len = data.len();
+        let response = unsafe { invoke(INVOKE_CALL_TAG, data.as_ptr(), len as u32) };
+        let (state_modified, res) = parse_response_code(response)?;
+        if state_modified {
+            // The state of the contract changed as a result of the call.
+            // So we refresh it.
+            if let Ok(new_state) = (&mut ContractState::open(())).get() {
```
Comment (Bargsteen) [inline concordium-std/src/impls.rs:588]: This does not work for `HasHost<ContractState>` which is what `low_level` methods will use.
Do we still intend to support the low_level interface in v1?

Comment (Bargsteen) [reply concordium-std/src/impls.rs:588]: One, less than ideal, option is: 
- Implement `Serialize` for `ContractState`, where `serial` writes nothing and `deserial` always resets the cursor to `0`.

File: concordium-std-derive/src/lib.rs [hunk 7]
```diff
@@ -663,18 +668,21 @@ fn receive_worker(attr: TokenStream, item: TokenStream) -> syn::Result<TokenStre
     );
 
     let mut out = if receive_attributes.optional.low_level {
```
[the remark this answers is not in the document]
Comment (Bargsteen) [reply concordium-std-derive/src/lib.rs:670]: I updated the counter example to see if it worked, and it almost did. The only problem is the `Serialize` requirement on the `HasHost` that I mention in a comment further below. 
I got it compiling after implementing a `Serial` + `Deserial` instance for `ContractState`, but I don't think these make much sense to have.

File: concordium-std/src/impls.rs [hunk 8]
```diff
@@ -419,12 +473,132 @@ impl<T: sealed::ContextType> HasCommonData for ExternContext<T> {
...
+                self.state = new_state;
+            } else {
+                return Err(InvokeError::Unknown);
```
[the remark this answers is not in the document]
Comment (abizjak) [reply concordium-std/src/impls.rs:591]: This error cannot happen.

I am not sure whether it is best to remove this variant, or mark the enum as non-completer, or non of these.

File: concordium-std-derive/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (abizjak) [reply concordium-std-derive/src/lib.rs:670]: Indeed it does not.

File: concordium-std/src/test_infrastructure.rs [hunk 9]
```diff
@@ -695,6 +655,259 @@ impl<T: AsRef<[u8]>> Seek for ContractStateTest<T> {
...
 }
 
+impl HasCallResponse for Cursor<Vec<u8>> {
+    fn size(&self) -> u32 { self.data.len() as u32 }
+}
+
+/// Holds a function used for mocking invocations of contracts with
+/// `invoke_contract`.
+pub struct MockFn<State>(
+    Box<dyn Fn(Parameter, Amount, &mut State, &mut Vec<u8>) -> InvokeResult<bool>>,
```
[the remark this answers is not in the document]
Comment (abizjak) [reply concordium-std/src/test_infrastructure.rs:665]: That'll be less ergonomic, and test infrastructure already uses the allocator, so this is fine.

Comment (abizjak) [inline concordium-std/src/test_infrastructure.rs:665]: Due to re-entrancy the balance of the contract can also be affected by this out-call. So there should probably be a way to allow users to specify that.

Comment (abizjak) [reply concordium-std/src/test_infrastructure.rs:665]: For example by adding another `&mut Amount` parameter.

File: concordium-std/src/prims.rs [hunk 10]
```diff
@@ -5,48 +5,62 @@
...
+    /// Invoke a host instruction. The arguments are
+    ///
+    /// - `tag`, which instruction to invoke
+    ///   - 0 for transfer to account
+    ///   - 1 for call to a contract
+    /// - `start`, pointer to the start of the invoke payload
+    /// - `length`, length of the payload
+    /// The response is encoded as follows.
+    /// - success is encoded as 0
+    /// - every failure has all bits of the first 3 bytes set
```
[the remark this answers is not in the document]
Comment (abizjak) [reply concordium-std/src/prims.rs:17]: Fixed.

File: concordium-std/src/impls.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (abizjak) [reply concordium-std/src/impls.rs:591]: I removed unknown.

File: concordium-std/src/test_infrastructure.rs (hunk 9 above)
Comment (abizjak) [reply concordium-std/src/test_infrastructure.rs:665]: Another thing,  failure can also return values, but this is not currently possible with this API.

File: concordium-std/src/types.rs [hunk 11]
```diff
@@ -6,10 +8,54 @@ pub struct ContractState {
...
+    pub(crate) i:                NonZeroU32,
+    pub(crate) current_position: u32,
+}
+
+/// A type representing the return value of contract init or receive method.
+pub struct ReturnValue {
     pub(crate) current_position: u32,
 }
 
+/// FIXME: Have two error types, one for transfers, one for calls.
```
[the remark this answers is not in the document]
Comment (abizjak) [reply concordium-std/src/types.rs:30]: Done.

File: concordium-std-derive/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (abizjak) [reply concordium-std-derive/src/lib.rs:670]: I am not sure which example is the counter example.

But I added handling of low_level now, and a low-level entrypoint to the `fib` example. I believe that works well.

**Comment by abizjak** [comment]
TODO: We need to update documentation of Reject and derive(Reject) to tell people that they can return values now.

File: concordium-std/src/test_infrastructure.rs (hunk 9 above)
Comment (abizjak) [reply concordium-std/src/test_infrastructure.rs]: I have addressed the latter now.

File: concordium-std/src/test_infrastructure.rs [hunk 12]
```diff
@@ -695,6 +655,313 @@ impl<T: AsRef<[u8]>> Seek for ContractStateTest<T> {
...
 
+impl HasCallResponse for Cursor<Vec<u8>> {
+    fn size(&self) -> u32 { self.data.len() as u32 }
+}
+
+/// Holds a function used for mocking invocations of contracts with
+/// `invoke_contract`.
+/// The provided closure may mutate its environment, for example to keep track
+/// of how many times the function was called during testing. This is the reason
+/// for `FnMut` bound.
```
Comment (Bargsteen) [inline concordium-std/src/test_infrastructure.rs:666]: This conflicts with the docs for `TestMockFn`.

Comment (Bargsteen) [reply concordium-std/src/test_infrastructure.rs:666]: Should we make it `Fn` for now? Or what are the awkward patterns for mutating the environment that you mention?

Comment (abizjak) [reply concordium-std/src/test_infrastructure.rs:666]: Let's keep it fnmut. If I missed updating documentation please update it.

**Pull request merged**